kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/CSVUtil.java

```diff
@@ -133,19 +134,23 @@ private static Schema inferSchemaInternal(String name, InputStream incoming,
     boolean[] nullable = new boolean[header.length];
     boolean[] empty = new boolean[header.length];
 
-    for (int processed = 0; processed < DEFAULT_INFER_LINES; processed += 1) {
+    for (int processed = 0; processed < props.linesToSample; processed += 1) {
       if (line == null) {
         break;
       }
 
       for (int i = 0; i < header.length; i += 1) {
         if (i < line.length) {
-          if (types[i] == null) {
-            types[i] = inferFieldType(line[i]);
+          Schema.Type prevType = types[i];
+          Schema.Type currType = inferFieldType(line[i]);
+          if (prevType == null) {
+            types[i] = currType;
             if (types[i] != null) {
-              // keep track of the value used
               values[i] = line[i];
             }
+          } else if (promoteType(prevType, currType)) {
```

Comment:
I think this would be better if you changed `promoteType` to `mergeTypes` (I think there might already be one somewhere. Maybe SchemaUtil or JsonUtil).